src/Microsoft.Framework.OptionsModel/ConfigurationBinder.cs

```diff
@@ -0,0 +1,253 @@
+        {
+            _startConfiguration = configuration;
+        }
+
+        public static void Bind(IConfiguration configuration, object model)
+        {
+            new ConfigurationBinder(configuration).Bind(model);
+        }
+
+        public void Bind(object model)
+        {
+            if (model == null)
+            {
+                return;
+            }
+
+            BindObject(model, _startConfiguration);
+        }
+
+        public TOptions Bind<TOptions>() where TOptions : new()
+        {
+            var model = new TOptions();
+            Bind(model);
+            return model;
+        }
+
+        private static void BindObject(object obj, IConfiguration configuration)
+        {
+            foreach (var property in GetAllProperties(obj.GetType().GetTypeInfo()))
+            {
+                BindProperty(property, obj, configuration);
+            }
+        }
+
+        private static void BindProperty(PropertyInfo property, object propertyOwner, IConfiguration configuration)
+        {
+            configuration = configuration.GetSubKey(property.Name);
+
+            if (property.GetMethod == null || !property.GetMethod.IsPublic)
+            {
+                // We don't support set only properties
+                return;
+            }
+
+            var propertyValue = property.GetValue(propertyOwner);
+            var hasPublicSetter = property.SetMethod != null && property.SetMethod.IsPublic;
+
+            if (propertyValue == null && !hasPublicSetter)
+            {
+                // Property doesn't have a value and we cannot set it so there is no
+                // point in going further down the graph
+                return;
+            }
+
+            propertyValue = BindType(property.PropertyType, propertyValue, configuration);
+
+            if (propertyValue != null && hasPublicSetter)
+            {
+                property.SetValue(propertyOwner, propertyValue);
+            }
+        }
+
+
+        private static object BindType(Type type, object typeInstance, IConfiguration configuration)
+        {
+            var configValue = configuration.Get(null);
+            var typeInfo = type.GetTypeInfo();
+
+            if (configValue != null)
+            {
+                // Leaf nodes are always reinitialized
+                return CreateValueFromConfiguration(type, configValue, configuration);
+            }
+            else
+            {
+                var subkeys = configuration.GetSubKeys();
+                if (subkeys.Count() != 0)
+                {
+                    if (typeInstance == null)
+                    {
+                        try
+                        {
+                            typeInstance = Activator.CreateInstance(type);
+                        }
+                        catch
+                        {
+                            // Cannot create the value so we cannot populate it
```

Comment:
Oh yes https://github.com/aspnet/Options/issues/55 is a great bug :smile: assigned to you :smile: I'm ok doing in a separate PR.